server/API.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	"net/http"
+)
+
+// UserIDHeaderKey - key of an http header where the user id is stored
+const UserIDHeaderKey = "Mattermost-User-ID"
+
+// APIError - returned error message for api errors
+type APIError struct {
+	Message    string `json:"message"`
+	StatusCode int    `json:"status_code"`
+}
+
+// writeAPIError writes api error as json in response
+func writeAPIError(w http.ResponseWriter, err *APIError) {
+	b, _ := json.Marshal(err)
+	w.WriteHeader(err.StatusCode)
+	_, _ = w.Write(b)
+}
+
+func respondWithJSON(w http.ResponseWriter, data interface{}) {
+	resp, _ := json.Marshal(data)
+	_, _ = w.Write(resp)
+}
+
+func writeSuccess(w http.ResponseWriter) {
+	_, _ = w.Write([]byte("{\"status\": \"OK\"}"))
+}
+
+// getUserIDFromRequest reads mattermost user ID from request
+func getUserIDFromRequest(r *http.Request) (string, error) {
+	userID := r.Header.Get(UserIDHeaderKey)
+	if userID == "" {
+		return "", errors.New("not authorized")
+	}
+	return userID, nil
+}
+
+// APICallHandler api call handler interface
+type APICallHandler func(p *Plugin, c *plugin.Context, w http.ResponseWriter, r *http.Request)
```

Comment:
Why are those functions which have `*Plugin` as a parameter and not methods of `Plugin`?